packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+  }
+}
+
+export function* render(value: unknown): IterableIterator<string> {
+  yield* renderValue(value, {customElementInstanceStack: []});
+}
+
+export function* renderValue(
+  value: unknown,
+  renderInfo: RenderInfo
+): IterableIterator<string> {
+  if (isRenderLightDirective(value)) {
+    // If a value was produced with renderLight(), we want to call and render
+    // the renderLight() method.
+    const instance = getLast(renderInfo.customElementInstanceStack);
+    if (instance !== undefined) {
+      yield* instance.renderLight(renderInfo);
+    }
+    value = null;
+  } else if (value != null && (value as DirectiveResult)._$litDirective$) {
+    const directive = (value as DirectiveResult)._$litDirective$;
+    value = new directive({type: NODE_PART}).render(
+      ...(value as DirectiveResult).values
+    );
+  }
+  if (value != null && (value as TemplateResult)._$litType$ !== undefined) {
+    yield `<!--lit-part ${digestForTemplateResult(value as TemplateResult)}-->`;
+    yield* renderTemplateResult(value as TemplateResult, renderInfo);
+  } else {
+    yield `<!--lit-part-->`;
+    if (
+      value === undefined ||
+      value === null ||
+      value === nothing ||
+      value === noChange
+    ) {
+      // yield nothing
+    } else if (Array.isArray(value)) {
+      for (const item of value) {
+        yield* renderValue(item, renderInfo);
+      }
+    } else {
+      yield escapeTextContent(String(value));
+    }
+  }
+  yield `<!--/lit-part-->`;
+}
+
+export function* renderTemplateResult(
+  result: TemplateResult,
+  renderInfo: RenderInfo
+): IterableIterator<string> {
+  // In order to render a TemplateResult we have to handle and stream out
+  // different parts of the result separately:
+  //   - Literal sections of the template
+  //   - Defined custom element within the literal sections
+  //   - Values in the result
+  //
+  // This means we can't just iterate through the template literals and values,
+  // we must parse and traverse the template's HTML. But we don't want to pay
+  // the cost of serializing the HTML node-by-node when we already have the
+  // template in string form. So we parse with location info turned on and use
+  // that to index into the HTML string generated by TemplateResult.getHTML().
+  // During the tree walk we will handle expression marker nodes and custom
+  // elements. For each we will record the offset of the node, and output the
+  // previous span of HTML.
+
+  const {ops, attrNames} = getTemplate(result);
+
+  /* The next value in result.values to render */
+  let partIndex = 0;
+  let attrIndex = 0;
+
+  for (const op of ops) {
+    switch (op.type) {
+      case 'text':
+        yield op.value;
+        break;
+      case 'node-part': {
+        const value = result.values[partIndex++];
+        yield* renderValue(value, renderInfo);
+        break;
+      }
+      case 'attribute-part': {
+        const name = attrNames[attrIndex++];
```

Comment:
This case is pretty big. Maybe factor into separate function for readability?

Reply:
And can we just skip this whole thing if the `value === noChange` and then remove the other checks for that?

Reply:
Done